Update docs for Leader Leases GA in v25.2 (#19587)
cockroachdb/docs: src/current/v25.2/architecture/replication-layer.md

A single node in the Raft group acts as the leaseholder, which is the only node that can serve reads or propose writes to the Raft group leader (both actions are received as `BatchRequests` from [`DistSender`]({% link {{ page.version.version }}/architecture/distribution-layer.md %}#distsender)).

CockroachDB ensures that the leaseholder is also the Raft group leader via the [Leader leases](#leader-leases) mechanism. This optimizes the speed of writes, and makes the cluster more robust against network partitions and node liveness failures.

When the leaseholder is sent a write request, a majority of the replica nodes must be able to communicate with each other to coordinate the write. This ensures that the most recent write is always available to subsequent reads.

If there is no leaseholder, any node receiving a request will attempt to become the leaseholder for the range.

> To extend its leases, each node must also remain the Raft leader as described in [Leader leases](#leader-leases). When a node disconnects, it stops updating its _store liveness_, causing the node to [lose all of its leases](#how-leases-are-transferred-from-a-dead-node).

Reviewer: ```suggestion
To extend its leases, each node must also remain the Raft leader, as described in [Leader leases](#leader-leases). When a node disconnects, it stops updating its _store liveness_, causing the node to [lose all of its leases](#how-leases-are-transferred-from-a-dead-node).
```

Author: ✅

> A table's meta and system ranges (detailed in the [distribution layer]({% link {{ page.version.version }}/architecture/distribution-layer.md %}#meta-ranges)) are treated as normal key-value data, and therefore have leases just like table data.

Reviewer: ```suggestion
A table's meta and system ranges (detailed in [Distribution Layer]({% link {{ page.version.version }}/architecture/distribution-layer.md %}#meta-ranges)) are treated as normal key-value data, and therefore have leases just like table data.
```

Author: ✅

CockroachDB is considered a CAP-Consistent (CP) system under the [CAP theorem](https://wikipedia.org/wiki/CAP_theorem). CockroachDB prioritizes data consistency, but also provides high-availability due to the coordination between leaseholders and replicas, minimizing the tradeoff between consistency and availability outlined by the CAP theorem. When you store data in CockroachDB, the data will be valid and anomaly-free, even in the event of a system error or power failure.

#### Co-location with Raft leadership

The range lease is always colocated with Raft leadership via the [Leader leases](#leader-leases) mechanism, except briefly during [lease transfers](#how-leases-are-transferred-from-a-dead-node). This reduces network round trips since the leaseholder receiving the requests can simply propose the Raft commands to itself, rather than communicating them to another node.

It also increases robustness against network partitions and outages due to liveness failures.

> For more information, see [Leader leases](#leader-leases).

Reviewer: ```suggestion
For more information, refer to [Leader leases](#leader-leases).
```

Author: ✅

CockroachDB is considered a CAP-Consistent (CP) system under the [CAP theorem](https://wikipedia.org/wiki/CAP_theorem). CockroachDB prioritizes data consistency, but also provides high-availability due to the coordination between leaseholders and replicas, minimizing the tradeoff between consistency and availability outlined by the CAP theorem. When you store data in CockroachDB, the data will be valid and anomaly-free, even in the event of a system error or power failure.

Because leases do not expire until a node disconnects from a cluster, leaseholders do not have to individually renew their own leases. Tying lease lifetimes to node liveness in this way lets us eliminate a substantial amount of traffic and Raft processing we would otherwise incur, while still tracking leases for every range.
#### Co-location with Raft leadership

> #### Expiration-based leases (meta and system ranges)

Author: @miraradeva @iskettaneh since per your other comments we still have both epoch- and expiration-based leases for some things, do you think i should refrain from deleting these sections, but just rewrite them?

Reviewer 2: So by default, we will have leader-leases (and expiration-based leases for some system ranges and for some operations like lease transfers). However, epoch leases are not there by default anymore (customers could use it if they want in 25.2).

A table's meta and system ranges (detailed in the [distribution layer]({% link {{ page.version.version }}/architecture/distribution-layer.md %}#meta-ranges)) are treated as normal key-value data, and therefore have leases just like table data.

Unlike table data, system ranges use expiration-based leases; expiration-based leases expire at a particular timestamp (typically after a few seconds). However, as long as a node continues proposing Raft commands, it continues to extend the expiration of its leases. If it doesn't, the next node containing a replica of the range that tries to read from or write to the range will become the leaseholder.

Expiration-based leases are used:

> - By some system ranges, as described above.

Reviewer: Can we cite the exact section this is referring to?

```suggestion
- By some system ranges, as described above.
```

Author: I realized we just explained this in the previous paragraph (where it says "Unlike table data, system ranges use expiration-based leases") so referring back to something so recent is probably not necessary

instead i've updated this last bit to read 

> Expiration-based leases are also used temporarily during operations like lease transfers, until the new Raft leader can be fortified based on store liveness as described in [Leader leases](#leader-leases).

> - Temporarily during some operations like lease transfers, until the new Raft leader can be fortified based on store liveness as described in [Leader leases](#leader-leases).

Reviewer: ```suggestion
- Temporarily during some operations like lease transfers, until the new Raft leader can be fortified based on store liveness, as described in [Leader leases](#leader-leases).
```

Author: ✅

- [Network partitions]({% link {{ page.version.version }}/cluster-setup-troubleshooting.md %}#network-partition) between a leaseholder and its followers heal in less than 20 seconds, since the leaseholder no longer needs to heartbeat a single node liveness range.
- Outages caused by liveness failures last less than 1 second, since liveness is now determined by a store-level detection mechanism, not a single node liveness range.
- Performance is equivalent (within less than 1%) to epoch-based leases on a 100 node cluster of 32 vCPU machines with 8 stores each.

{% include {{ page.version.version }}/leader-leases-node-heartbeat-use-cases.md %}

### How leases are transferred from a dead node

When a cluster needs to access a range on a leaseholder node that is dead, the lease must be transferred to a healthy node. The process is as follows:

1. Detection of Node Failure: The _store liveness_ mechanism described in [Leader leases](#leader-leases) detects node failures through its store-wide heartbeating process. If a node becomes unresponsive, its store liveness support is withdrawn, marking it as unavailable.

> 1. Raft Leadership Election: A Raft election is initiated to establish a new leader for the range. This step is necessary because lease acquisition can only occur on the Raft leader. The election process includes a store liveness component to fortify the new leader as described in [Leader leases](#leader-leases).

Reviewer: ```suggestion
1. Raft Leadership Election: A Raft election is initiated to establish a new leader for the range. This step is necessary because lease acquisition can only occur on the Raft leader. The election process includes a store liveness component to fortify the new leader, as described in [Leader leases](#leader-leases).
```

Author: ✅
